awebox/mdl/dynamics_components_dir/mass.py

```diff
@@ -0,0 +1,213 @@
+#
+#    awebox is free software; you can redistribute it and/or
+#    modify it under the terms of the GNU Lesser General Public
+#    License as published by the Free Software Foundation; either
+#    version 3 of the License, or (at your option) any later version.
+#
+#    awebox is distributed in the hope that it will be useful,
+#    but WITHOUT ANY WARRANTY; without even the implied warranty of
+#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+#    Lesser General Public License for more details.
+#
+#    You should have received a copy of the GNU Lesser General Public
+#    License along with awebox; if not, write to the Free Software Foundation,
+#    Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+#
+#
+'''
+node mass computation
+_python-3.5 / casadi-3.4.5
+- edited: rachel leuthold, jochem de schutter alu-fr 2017-20
+'''
+
+import casadi.tools as cas
+
+import awebox.tools.vector_operations as vect_op
+import awebox.tools.struct_operations as struct_op
+import awebox.tools.print_operations as print_op
+
+import awebox.mdl.dynamics_components_dir.tether as tether
+
+from awebox.logger.logger import Logger as awelogger
+
+
+def generate_m_nodes(options, variables, outputs, parameters, architecture):
+    # system architecture (see zanon2013a)
+    number_of_nodes = architecture.number_of_nodes
+    parent_map = architecture.parent_map
+
+    node_masses = generate_mass_dictionary_for_all_nodes(options, variables, parameters, architecture, 'vals')
+
+    # save some space in the outputs for the node masses
+    if 'masses' not in list(outputs.keys()):
+        outputs['masses'] = {}
+
+    # save the each node's responsible mass into the outputs
+    for node in range(1, number_of_nodes):
+        parent = parent_map[node]
+        outputs['masses']['m' + str(node) + str(parent)] = node_masses['m' + str(node) + str(parent)]
+
+    if options['tether']['use_wound_tether']:
+        outputs['masses']['m00'] = node_masses['m00']
+
+    return node_masses, outputs
+
+
+
+def generate_m_nodes_scaling(options, variables, outputs, parameters, architecture):
+    # system architecture (see zanon2013a)
+    number_of_nodes = architecture.number_of_nodes
+    parent_map = architecture.parent_map
+
+    node_masses_scaling = generate_mass_dictionary_for_all_nodes(options, variables, parameters, architecture,
+                                                                 'scaling')
+
+    # this will be used to scale the forces,
+    # so we need to repeat the scaling forces 3x, once per dimension for force.
+    # and, only on those nodes for which we will construct dynamics equations via. Lagrangian mechanics
+
+    node_masses_scaling_stacked = []
+    for node in range(1, number_of_nodes):
+        parent = parent_map[node]
+        mass = node_masses_scaling['m' + str(node) + str(parent)]
+
+        # if node == 1:
+        #     mass += node_masses_scaling['groundstation']
```

Comment:
is this relevant? if not, it can be deleted

Reply:
thank you. ground-station mass was already added in node_masses_scaling...